File: Trees/DFS/1120. Maximum Average Subtree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maximumAverageSubtree = function(root) 
{
    let maxAverage = 0;

    // Helper function to calculate sum and count of nodes in a subtree
    const dfs = (node) => 
    {
        if (!node) 
            return [0, 0]; // Base case: null node
        
        // Postorder traversal: compute for left and right subtrees first
        const [leftSum, leftCount] = dfs(node.left);
        const [rightSum, rightCount] = dfs(node.right);

        // Compute current subtree's sum and count
        const currentSum = leftSum + rightSum + node.val;
        const currentCount = leftCount + rightCount + 1;

        // Update the maximum average
        const currentAverage = currentSum / currentCount;
        maxAverage = Math.max(maxAverage, currentAverage);

        // Return the sum and count for the current subtree
        return [currentSum, currentCount];
    };

    dfs(root);
    return maxAverage;
};

File: Trees/DFS/1120. Maximum Average Subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maximumAverageSubtree(self, root: Optional[TreeNode]) -> float:
        self.average = 0

        def dfs(node):
            if not node:
                return (0, 0) #(sum, count)

            leftSum, leftCount = dfs(node.left)
            rightSum, rightCount = dfs(node.right)

            totalSum = leftSum + rightSum + node.val
            totalCount = leftCount + rightCount + 1

            self.average = max(totalSum / totalCount, self.average)

            return (totalSum, totalCount)
        

        totalSum, count = dfs(root)
        return self.average